connector/consumer.go

```diff
@@ -0,0 +1,132 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connector // import "go.opentelemetry.io/collector/connector"
+
+import (
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/fanoutconsumer"
+)
+
+// TracesConsumer feeds the first consumer.Traces in each of the specified pipelines.
+// Unlike other consumers, the connector consumer will expose its underlying consumers.
+type TracesConsumer interface {
+	consumer.Traces
+	Consumers() map[component.ID]consumer.Traces
+}
+
+var _ TracesConsumer = (*tracesConsumer)(nil)
+
+type tracesConsumer struct {
+	consumer.ConsumeTracesFunc
+	consumers map[component.ID]consumer.Traces
+}
+
+func NewTracesConsumer(cm map[component.ID]consumer.Traces) TracesConsumer {
```

Comment:
I would definitely make this internal. Find a place somewhere in service/internal so no need on exposing the fanout components.

Reply:
> For example, the current routing processor allows to send to multiple routes per entry in the routing table. This is ok in the current implementation because we route directly to exporters, which we assume do not mutate the data.

That is a wrong assumption.

We can expose fanout, but still don't have to expose the implementation of the `TracesRouter`.

Reply:
The assumption may be bad, but it is a good example how easy it is to ignore mutability because the current routing processor does not account for it.

Exposing fanout would make it _easier_ to do the right thing in such cases, but we will still have to enforce it better than we did with the current routing processor. I'm fine with this, but still think https://github.com/open-telemetry/opentelemetry-collector/pull/7179 is a better solution because it does not allow connector authors to make for such errors.

Reply:
Related to https://github.com/open-telemetry/opentelemetry-collector/pull/7179 we agreed that is hard to use for the fact that needs to know IDs in advance.

Reply:
> Related to #7179 we agreed that is hard to use for the fact that needs to know IDs in advance.

Yeah, that's true. I think we can adapt the interface to provide this though.